### PR TITLE
add sign-invoice command

### PR DESCRIPTION
**Pull request by technosophos**
This is the third step toward implementing signing and verification. It uses a key created with `create-key` and signs an invoice.

Signed-off-by: Matt Butcher <matt.butcher@microsoft.com>

### CODE REVIEW

File: bin/client/main.rs [hunk 1]
```diff
@@ -91,6 +92,37 @@ async fn main() -> std::result::Result<(), ClientError> {
...
+            // Role
+            let role = if let Some(r) = sign_opts.role {
+                role_from_name(r)?
+            } else {
+                SignatureRole::Creator
+            };
+            // Keyfile
+            let keyfile = sign_opts.secret_file.unwrap_or_else(|| {
+                default_config_dir()
+                    .unwrap_or_else(|| PathBuf::from("."))
```
Comment (thomastaylor312) [inline bin/client/main.rs:105]: Nit: I think you can just do `PathBuf::from("secret_keys.toml")` here

File: src/invoice/mod.rs [hunk 2]
```diff
@@ -130,8 +130,19 @@ impl Invoice {
         &mut self,
         signer_name: String,
         signer_role: SignatureRole,
-        key: Keypair,
+        key: &Keypair,
     ) -> Result<(), SignatureError> {
+        // The spec says it is illegal for the a single key to sign the same invoice
+        // more than once.
+        let encoded_key = base64::encode(key.public.to_bytes());
+        if let Some(sigs) = self.signature.as_ref() {
```
Comment (thomastaylor312) [inline src/invoice/mod.rs:138]: This `if let` and `for` + `if` block could be reduced to something close to this:

```rust
if self.signature.as_ref().map(|sigs|  sigs.iter().any(|s| s.key == encoded_key) ).unwrap_or(false) {
    return Err(SignatureError::DuplicateSignature)
}
```

**Review by thomastaylor312 — APPROVED** [review]
Some minor nits. Anything else I figured we'd fix/refactor in our pairing session

File: bin/client/main.rs (hunk 1 above)
Comment (technosophos) [reply bin/client/main.rs:105]: I don't think so... that would give me `secret_keys.toml/secret_keys.toml`. The idea is to get the default config dir or else use `.`, and _then_ append `secret_keys.toml`. So if config dir is `/foo/bar`, the result should be `/foo/bar/secret_keys.toml`.

Or maybe I'm misunderstanding something about what you are suggesting?

Comment (thomastaylor312) [reply bin/client/main.rs:105]: Oh right, you are joining on the result of the operation. Ignore me

File: src/invoice/mod.rs (hunk 2 above)
Comment (technosophos) [reply src/invoice/mod.rs:138]: This part is already re-written for the next PR... so gonna leave this for now and just merge it in the next PR

**Pull request merged**